File: backend/app/models/schemas.py
"""
Copyright (c) 2025, elk-MCP Project.
All rights reserved.
"""

from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field, constr


class Pagination(BaseModel):
    page: int = Field(1, ge=1, le=100000)
    page_size: int = Field(50, ge=1, le=200)


class TimeRange(BaseModel):
    start: constr(strip_whitespace=True)
    end: constr(strip_whitespace=True)


class LogQueryFilters(BaseModel):
    level: Optional[List[str]] = None
    service: Optional[List[str]] = None
    keyword: Optional[str] = None


class SortSpec(BaseModel):
    field: str = Field("timestamp")
    order: str = Field("desc")


class LogQueryRequest(BaseModel):
    tenant_id: str
    pagination: Pagination
    time_range: TimeRange
    filters: LogQueryFilters = Field(default_factory=LogQueryFilters)
    sort: SortSpec = Field(default_factory=SortSpec)
    # Dynamic index selection
    index_keyword: Optional[str] = None
    use_regex: Optional[bool] = False
    override_indexes: Optional[List[str]] = None


class AlertRuleRef(BaseModel):
    id: str
    severity: Optional[str] = None


class AlertsQueryRequest(BaseModel):
    tenant_id: str
    time_range: TimeRange
    severity: Optional[List[str]] = None
    rules: Optional[List[AlertRuleRef]] = None


class StatsRequest(BaseModel):
    tenant_id: str
    time_range: TimeRange
    group_by: str = Field(..., pattern=r"^(service|level|host)$")


class StandardLog(BaseModel):
    timestamp: str
    level: Optional[str]
    message: str
    service: Optional[str]
    tenant_id: str
    host: Optional[str]
    extra: Dict[str, Any] = Field(default_factory=dict)


class QueryResponse(BaseModel):
    code: int
    i18n_key: str
    data: Dict[str, Any]


class IndexDiscoveryConfig(BaseModel):
    enabled: Optional[bool] = None
    interval_seconds: Optional[int] = Field(default=None, ge=5, le=3600)
    include_patterns: Optional[List[str]] = None
    exclude_patterns: Optional[List[str]] = None

File: schemas-ts/src/schemas.ts
/*
Copyright (c) 2025, elk-MCP Project.
All rights reserved.
*/

import { z } from 'zod';

export const Pagination = z.object({
  page: z.number().int().min(1).max(100000),
  page_size: z.number().int().min(1).max(200),
});

export const TimeRange = z.object({
  start: z.string().min(1),
  end: z.string().min(1),
});

export const LogQueryFilters = z.object({
  level: z.array(z.string()).optional(),
  service: z.array(z.string()).optional(),
  keyword: z.string().optional(),
});

export const SortSpec = z.object({
  field: z.enum(['timestamp', '_score']).default('timestamp'),
  order: z.enum(['asc', 'desc']).default('desc'),
});

export const LogQueryRequest = z.object({
  tenant_id: z.string().min(1),
  pagination: Pagination,
  time_range: TimeRange,
  filters: LogQueryFilters.default({}),
  sort: SortSpec.default({ field: 'timestamp', order: 'desc' }),
  mode: z.enum(['page', 'cursor']).default('page'),
  cursor_after: z.array(z.union([z.string(), z.number()])).optional(),
});

export const AlertRuleRef = z.object({ id: z.string().min(1), severity: z.string().optional() });

export const AlertsQueryRequest = z.object({
  tenant_id: z.string().min(1),
  time_range: TimeRange,
  severity: z.array(z.enum(['low', 'medium', 'high'])).optional(),
  rules: z.array(AlertRuleRef).optional(),
});

export const StatsRequest = z.object({
  tenant_id: z.string().min(1),
  time_range: TimeRange,
  group_by: z.enum(['service', 'level', 'host']),
});

export type TLogQueryRequest = z.infer<typeof LogQueryRequest>;
export type TAlertsQueryRequest = z.infer<typeof AlertsQueryRequest>;
export type TStatsRequest = z.infer<typeof StatsRequest>;

